analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs

```diff
@@ -287,5 +288,246 @@ public void DeclarationPattern_Var_Deconstruction_DoesNotSetConstraint()
+        [DataRow("stringUnknown", "{ Length: > 10 } and { Length: < 100 }", null)]
+
+        [DataRow("objectNull", "null or not { }", true)]
+        [DataRow("objectNotNull", "null or not { }", false)]
+        [DataRow("objectUnknown", "null or not { }", null)]
+        [DataRow("objectNull", "null or { }", true)]
+        [DataRow("objectNotNull", "null or { }", true)]
+        [DataRow("objectUnknown", "null or { }", null)]  // FN. Matches always.
+        [DataRow("stringNull", "{ Length: 0 } or not null", false)]
+        [DataRow("stringNotNull", "{ Length: 0 } or not null", true)]
+        [DataRow("stringNotNull", "{ Length: 0 } or null", null)]
+        [DataRow("stringUnknown", "{ Length: 0 } or not null", null)]
+        [DataRow("stringNull", "not null or { Length: 0 }", false)]
+        [DataRow("stringNotNull", "not null or { Length: 0 }", true)]
+        [DataRow("stringUnknown", "not null or { Length: 0 }", null)]
+        [DataRow("stringNull", "{ Length: > 10 } or { Length: < 100 }", false)]
+        [DataRow("stringNotNull", "{ Length: > 10 } or { Length: < 100 }", null)]
+        [DataRow("stringUnknown", "{ Length: > 10 } or { Length: < 100 }", null)]
+        public void AndOrPatternsSetBoolConstraint(string variableName, string isPattern, bool? expectedBoolConstraint)
+        {
+            const string variableDeclarations = @"
+var objectNotNull = new object();
+var objectNull = (object)null;
+var objectUnknown = Unknown<object>();
+var stringNotNull = new string('c', 1);  // Make sure, we learn 'stringNotNull is not null'
+var stringNull = (string)null;
+var stringUnknown = Unknown<string>();
+var exceptionNotNull = new Exception();
+";
+            ValidateSetBoolConstraint(variableDeclarations, variableName, isPattern, OperationKindEx.BinaryPattern, expectedBoolConstraint);
+        }
+
+        private static void ValidateSetBoolConstraint(string variableDeclarations,
+                                                      string variableName,
+                                                      string isPattern,
+                                                      OperationKind expectedOperation,
+                                                      bool? expectedBoolConstraint)
+            => ValidateSetBoolConstraint(additionalTypes: string.Empty, variableDeclarations, variableName, isPattern, expectedOperation, expectedBoolConstraint);
+
+        private static void ValidateSetBoolConstraint(string additionalTypes,
+                                                      string variableDeclarations,
+                                                      string variableName,
+                                                      string isPattern,
+                                                      OperationKind expectedOperation,
+                                                      bool? expectedBoolConstraint)
+        {
+            var code = @$"
+{additionalTypes}
```

Comment:
This should be at the end, after `Main`. Because CFG will be generated for the first method. Having it at the beginning can cause a random helper method from `additionalTypes` to be analyzed

Reply:
Good point. I hit that problem before.